### PR TITLE
Lnx-103 implement goto action

### PR DESCRIPTION
**Pull request by Krystian030**
- GoTo action has been implemented 
- To perform the aciton, specify `goto(Vector(x,y))`

### CODE REVIEW

File: src/runtime.py [hunk 1]
```diff
@@ -15,22 +17,37 @@ def execution_runtime(pipe: AioConnection, object_id: int):
     from lynx.common.vector import Vector
     from lynx.common.enums import Direction
 
-
     scene_serialized = pipe.recv()
     scene: Scene = Scene.deserialize(scene_serialized)
 
     def send(action: Action):
+        nonlocal scene
```
Comment (EasyJB) [inline src/runtime.py:33]: Is it needed?

File: src/runtime.py [hunk 2]
```diff
@@ -15,22 +17,37 @@ def execution_runtime(pipe: AioConnection, object_id: int):
...
+        agent = scene.get_object_by_id(object_id)
+        objects_on_target_position = scene.get_objects_by_position(target_position)
+        for object in objects_on_target_position:
+            if 'walkable' not in object.tags:
+                return
+
+        path = AStarAlgorithm(agent.position, target_position).get_path(scene)
+        for vector in path:
+            move(vector)
```
Comment (EasyJB) [inline src/runtime.py]: Wouldn't it be better to put `send()` here? I mean wouldn't it be clearer?

File: src/algorithms/a_star_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
+from dataclasses import dataclass
+from typing import List, Tuple
+
+from lynx.common.vector import Vector
+
+
+class Node:
+    def __init__(self, parent=None, position=None):
+        self.parent = parent
```
Comment (EasyJB) [inline src/algorithms/a_star_algorithm.py]: What is parent used to?

File: src/algorithms/a_star_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+class Node:
+    def __init__(self, parent=None, position=None):
+        self.parent = parent
+        self.position = position
+
+        self.g = 0
+        self.h = 0
+        self.f = 0
+
```
Comment (EasyJB) [inline src/algorithms/a_star_algorithm.py]: Aren't there better names, to make it readable

**Review by EasyJB — CHANGES_REQUESTED** [review]
Imo the a_star_algorithm.py looks kinda strange, because it's already implemented algorith, so I guess it's normal. There are some places in which I would change names, but maybe they are typical for algorith, so idk.

File: src/algorithms/a_star_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
+from dataclasses import dataclass
+from typing import List, Tuple
+
+from lynx.common.vector import Vector
+
+
+class Node:
```
Comment (filip-szweda) [inline src/algorithms/a_star_algorithm.py]: If `Node` is only ever used in `AStarAlgorithm`, you can declare it inside the class like:
```
@dataclass
class AStarAlgorithm:
    """
    Get the shortest path from start to end
    """
    class Node:
        ...
    ...
```

File: src/algorithms/a_star_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
+from dataclasses import dataclass
+from typing import List, Tuple
+
+from lynx.common.vector import Vector
+
+
+class Node:
+    def __init__(self, parent=None, position=None):
```
Comment (filip-szweda) [inline src/algorithms/a_star_algorithm.py]: Missing return type hint

File: src/algorithms/a_star_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        self.parent = parent
+        self.position = position
+
+        self.g = 0
+        self.h = 0
+        self.f = 0
+
+    def __eq__(self, other):
+        return self.position == other.position
```
Comment (filip-szweda) [inline src/algorithms/a_star_algorithm.py]: Are you sure that we should only compare `position` and not `parent`, `g`, `h`, `f`?

File: src/algorithms/a_star_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        closed_list = []
+
+        # add the start node
+        open_list.append(start_node)
+
+        # loop until you find the end
+        while len(open_list) > 0:
+            # get the current node
+            current_node = open_list[0]
+            current_index = 0
```
Comment (filip-szweda) [inline src/algorithms/a_star_algorithm.py]:
```
current_node_index = 0
current_node = open_list[current_node_index]
```

File: src/algorithms/a_star_algorithm.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            for index, item in enumerate(open_list):
```
Comment (filip-szweda) [inline src/algorithms/a_star_algorithm.py]:
```
for open_node_index, open_node in enumerate(open_list):
```

File: src/algorithms/a_star_algorithm.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+    @staticmethod
+    def transform_to_unit_vectors(positions: List[Tuple]) -> List[Vector]:
+        """
+        Transform list of tuples to list of vectors from start to end
+        :param positions: list of tuples
+        :return: list of unit vectors from start to end
+        """
+        vectors = []
+        for i in range(len(positions) - 1):
```
Comment (filip-szweda) [inline src/algorithms/a_star_algorithm.py]: You can use `enumerate` here

File: src/algorithms/a_star_algorithm.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+@dataclass
+class AStarAlgorithm:
+    """
+    Get the shortest path from start to end
+    """
+    start: Vector = None
+    end: Vector = None
+
+    def get_path(self, scene: 'Scene') -> List[Vector]:
```
Comment (filip-szweda) [inline src/algorithms/a_star_algorithm.py:27]: There's a HUGE opportunity to use helper functions here.
Consider it a rule of thumb, that if you need to write a comment for a fragment of code, this fragment of code can become a function.

For instance, this:
```
# found the goal
            if current_node == end_node:
                path = []
                current = current_node
                while current is not None:
                    path.append(current.position)
                    current = current.parent
                return self.transform_to_unit_vectors(path[::-1])
```
can become this:
```
if path := is_goal_found(current_node, end_node):
    return path
```

**Review by filip-szweda — CHANGES_REQUESTED** [review]

File: src/runtime.py (hunk 1 above)
Comment (Krystian030) [reply src/runtime.py:33]: yep, in goto I use `send` many times, so after first execution the scene is changed but after that it's not, so this is the reason why nonlocal is needed here

File: src/algorithms/a_star_algorithm.py (hunk 3 above)
Comment (Krystian030) [reply src/algorithms/a_star_algorithm.py]: To create final path in 63 line you change current to current.parent.

File: src/algorithms/a_star_algorithm.py (hunk 4 above)
Comment (Krystian030) [reply src/algorithms/a_star_algorithm.py]: Done

Comment (filip-szweda) [reply src/algorithms/a_star_algorithm.py]: Done, but not commited @Krystian030

File: src/algorithms/a_star_algorithm.py (hunk 7 above)
Comment (Krystian030) [reply src/algorithms/a_star_algorithm.py]: Yep, we should only compare position

File: src/algorithms/a_star_algorithm.py (hunk 6 above)
Comment (Krystian030) [reply src/algorithms/a_star_algorithm.py]: But I'm not returning anything

File: src/algorithms/a_star_algorithm.py (hunk 11 above)
Comment (Krystian030) [reply src/algorithms/a_star_algorithm.py:27]: Okay, I will keep this in mind in the future, for now I would leave it as is. On the other hand:
```
if path := is_goal_found(current_node, end_node):
    return path
```
`is_goal_found` suggests that a boolean value will be returned

File: src/runtime.py [hunk 12]
```diff
@@ -9,20 +13,49 @@ def execution_runtime(pipe: AioConnection, object_id: int):
...
+        agent = scene.get_object_by_id(object_id)
+        if agent.position == target_position:
+            send(MessageLog(object_id, f"I am already on target position"))
+
+        objects_on_target_position = scene.get_objects_by_position(target_position)
+        for object in objects_on_target_position:
+            if 'walkable' not in object.tags:
+                send(MessageLog(object_id, f"Target position: {target_position} is not walkable"))
+
+        path = AStarAlgorithm(agent.position, target_position).get_path(scene)
```
Comment (alpinus4) [inline src/runtime.py]: it's calculated once, and then we execute all the moves.
What if something blocks the way?
It has to be handled

**Review by alpinus4 — CHANGES_REQUESTED** [review]

File: src/runtime.py (hunk 12 above)
Comment (Krystian030) [reply src/runtime.py]: Done

File: src/runtime.py [hunk 13]
```diff
@@ -9,20 +13,66 @@ def execution_runtime(pipe: AioConnection, object_id: int):
...
+
+        objects_on_target_position = scene.get_objects_by_position(target_position)
+        for object in objects_on_target_position:
+            if 'walkable' not in object.tags:
+                send(MessageLog(object_id, f"Target position: {target_position} is not walkable"))
+                return
+
+        while True:
+            path = AStarAlgorithm(agent.position, target_position).get_path(scene)
+            obstacle_found = False
```
Comment (alpinus4) [inline src/runtime.py:57]: use `satisfies_requirements()`

**Review by alpinus4 — CHANGES_REQUESTED** [review]

**Review by EasyJB — APPROVED** [review]
LGTM

**Review by alpinus4 — APPROVED** [review]
Looks good :)

**Review by filip-szweda — APPROVED** [review]

**Pull request merged**